### ISSUE TITLE
SVG images and night mode - can we add CSS definition? #a11y issue

### ISSUE THREAD

**Issue by VeryGoodErotica** [w3c/epub-specs#1304]
I do not have a lot of hardware ePub readers to test with (I'm financially poor) but this is an issue with iBooks on 3rd gen iPad so I imagine it may be an issue elsewhere.

When using a text string in SVG where the SVG code is part of the xhtml itself (the only reliable way to use CSS to specify the font face), when iBooks is in night mode it disregards the color of the text (when specified) in the CSS and changes the color of the text so that it would have high contrast against a black background.

However the background color of the SVG itself is not changed.

While the text is usually still readable for people with good vision, it often results in text that does not meet the WCAG contrast threshold.

If we could have a CSS selector added to the spec for specifying the color of text when in night mode, then we could use it in cases like SVG where the background color does not change in night mode and therefore the color of the text in night mode should also not change.

Alternatively it could be specified that text nodes can not have their color changed in night mode, but that would create a similar problem when the SVG has a transparent background.

**Comment by dauwhe** [comment]
This sounds like a bug in iBooks rather than an issue with the EPUB specification. 

Note that the CSSWG is developing a media query to determine if the user prefers a color theme. See https://drafts.csswg.org/mediaqueries-5/#descdef-media-prefers-color-scheme

**Comment by VeryGoodErotica** [comment]
It's a difficult issue because if an SVG has a transparent background then changing the color of SVG text nodes is appropriate.

So either the reader has to be able to determine if the text is on a transparent background or a CSS selector for night mode is needed to turn off text color replacement in night mode when the SVG has a non-transparent background.

**Comment by Doktorchen** [comment]
As far as I understand, you have an SVG fragment within the XHTML document?
In this case all the specifities and selectors apply to SVG elements in the same way as to XHTML elements, you only have to select them properly.
But for colors you typically use the attributes fill, stroke etc; the attribute color applies only indirectly.

Additionally you can use the keyword currentColor for example for attributes like stroke and fill to specify the value of the color attribute.
If you need a background-color you have to add a large element like a circle as background in SVG 1.1 (SVG tiny 1.2 has itself background-color, but this may only apply for EPUB 3.2, because this version does specify, that SVG 1.1 has to be used, but maybe ebubcheck may still assume this).
If you use the class attribute or id attribute to this background element, you can apply easiliy another fill or stroke etc as well to this element with CSS selectors.

If a complete SVG document is referenced with elements like img or object, no properties are inherited from the embedding document.
You can use XML stylesheet processing instructions to apply CSS documents to these SVG documents as well and within the defs element you can add a style element with CSS in it in a similar way as in XHTML.

But XML processing instructions have no class attribute, therefore here no option to indicate something like day and night mode.
The SVG style element has no class attribute as well, but the title attribute, therefore possible to select here as with XML instructions alternative styles manually, but no EPUB specific indication of day and night mode.

**Comment by VeryGoodErotica** [comment]
I tried using
fill="black" in the text node to prevent it from changing color and that had no effect, iBooks still changed the color. And it probably should even with fill="black" because you can use colors for visual effect yet still have a transparent background requiring they be changed when nightmode is active. But when there is the background is not transparent, then the color of text nodes should not be changed - even if not specified.

There are ePub specific CSS instructions like `page-break-before: always;` so something like `night-mode: disable;` might do the trick, telling ePub viewers not to apply night mode style to anything in that selector class (whether an SVG node or an XHTML node)

**Comment by VeryGoodErotica** [comment]
I can not upload images here so here is a tweet with screen caps demonstrating the problem:

https://twitter.com/EroticaGood/status/1190591806628515840

**Comment by Doktorchen** [comment]
Hmm, ok, fill="black" is only a presentation attribute with low priority.
But would be really strange, if this user agent iBooks had a user-agent-stylesheet setting the fill property for text elements on specific values.

Unfortunately I have no access to such an iBooks viewer, but I did systematic EPUB tests on a larger number of other viewers (mainly browser extensions).

However, you can increase the specifity by adding fragmentidentifiers id to the root svg and the related text element.
Within your stylesheet applicable for a test SVG content document  you can try something (nasty) like this for a test:
:root:not(text)#mysvg text:not(svg)#mytext {fill: red}
For SVG fragments within XHTML you still can test something like:
svg#mysvg text#mytext,  svg#mysvg text:not(svg)#mytext {fill: red}
respectively:
html > body#someid svg#mysvg text#mytext {fill: red}
:root:not(body) > body:not(html) svg#mysvg text:not(svg)#mytext {fill: red}
All of them have a high specifity due to the combination of the selectors.

If the text is still not red but black, this might be an indication, that CSS interpretation within this iBooks is completely borked, because it ignores basic CSS rules about specifities and priorities of selectors and stylesheets from different sources.
This happens in some other EPUB viewers as well for a few XHTML elements, but would be new for SVG elements as well.

In general for day or night mode you provide different stylesheets for your documents.
In XHTML this looks like this:

    <link href="night.css" type="text/css" title="night" rel="Stylesheet" class="night"/>
    <link href="day.css" type="text/css" title="day" rel="Alternate Stylesheet" class="day"/>
    <link href="none.css" type="text/css" title="none" rel="Alternate Stylesheet" />
    <link href="always.css" type="text/css" rel="Stylesheet" />

Here always.css applies, whatever is selected by the audience or the viewer, except the interpretation of CSS is switched off by the audience.
none.css can be selected as alternative stylesheet additionally, because the related link element has a title attribute.
Similar for the day and night styles. The viewer may select one of them automatically, but still the audience has the choice to switch.

Therefore, it something has to apply always, if CSS interpretation is active at all, note it in always.css in this example.
Alternatively of course you can note the same selector rules both in the day and night styles, if it applies to both.

If it turns out, that the iBooks viewer is borked, it is far more meaningful to send a bug report about the issue to the developers of this viewers as to change EPUB to care about bugs in viewers.

However, what remains: There is no method to provide day and night styles for SVG content documents, this is a gap in the specification of these day and night styles - maybe anyway not the best idea to use the class attribute for such a purpose ;o)

**Comment by VeryGoodErotica** [comment]
Thank you for the very detailed explanation. Brings up another question unrelated - do the alternate style sheets work in modern ePub readers?

Right now on first page after title page, if the viewer supports JS with localStorage, I ask the user if they prefer bold text where italic is often used. If they do, then I store that in localStorage and every page loads an alternate stylesheet that overrides defaults for things like em and dfn (and my even my defined class for the math equation) but if more modern ePub viewers allow selection of alternate style sheets and it is obvious to users how to invoke it, that's clearly better. Seems a lot of ePub 3.2 spec though isn't actually implemented though by a lot of viewers.

**Comment by Doktorchen** [comment]
At least extensions for Mozilla/Gecko like browser have an option so switch between alternative styles provided by the author.
Unfortunately one has to select it again for each document again.
The same applies for the browser Opera (old versions using Presto, but there are not many EPUB extensions for this program).

Konqueror (KHTML, not WebKit/Webengine/Blink variant; KDE) has a complete implementation for the selection with alternative stylesheets, the selected style remains after a link to another document using stylesheets with the same names.
But I know no application or extension, that uses KHTML/Konqueror to display EPUBs, the audience would have to unzip first to get it work.

However, if a viewer/reader has at least a selection for day/night view, it has to select the related stylesheet, if provided by authors, this has priority over user-agent-stylesheets.

To select preferred user styling CSS require from viewers/readers to provide an option/interface to select a user-stylesheet.
However, this is often not easy to find or not implemented at all, therefore another bug or gap in CSS support similar to the missing option to select between alternative stylesheets.
But several viewers have options to select some basic styling features, maybe including bold or italic font variants, font selection at all, this applies in most cases at least for the preferred font-size or minimal font-size, therefore surely beyond alternative stylesheets not the task for authors to care about this with javaScript.

The same applies for accessibility problems due to bugs in a specific viewer: It is not the task of authors to test all viewers and to work around their bugs.

However, there is a declarative option (not recommended by me) to select substyling with a CSS selector combination like
#t1:target ~ #style *,  #t1:target ~ #main
In the XHTML document you have something like
&lt;body>
&lt;div class="top" id="t1"/>&lt;div class="top" id="t2"/> ...
&lt;nav id="style">
&lt;h1 title="select">Style&lt;/h1> 
&lt;ol>
&lt;li>&lt;a href="#t1">1&lt;/a>&lt;/li>
&lt;li>&lt;a href="#t2">2&lt;/a>&lt;/li>
...
&lt;/ol>
&lt;/nav>
&lt;main id="main">

Content ...

&lt;/main>
&lt;/body>

But this has to be selected again for each document again, requires therefore a style selection list at the beginning of each document.
And of course the document has not other links to fragment identifiers on the same page.
Obviously using a similar list for links to other documents it is possible as well to get the selection to the next page.
Unfortunately the idea with parameters added to IRIs is not realised, once there was this idea for SVG.
But because you can have this with get parameters in web projects anyway with script languages like PHP, there seems to be not much interest to enable this as well declaratively for static projects including EPUB content.
Once I tried this, but it is maybe only meaningful for very basic decorations, else the stylesheet gets pretty huge.
Would be much more useful to have the bugs/gaps in viewers fixed to get this functionality work as intended by CSS recommendations.
But they never fixed it in browsers like webKit/Blink or Microsoft browsers, therefore a low chance that this will be done for EPUB readers with much more basic bugs and problems.

Once I suggested to note a book wide stylesheet selection already within the OPF-file for EPUB, but there was no interest to specify this - maybe especially because quite often still EPUB 2 is used, not EPUB 3, therefore not much interest in EPUB specific features.

If developers of programs do not fix bugs/gaps, authors cannot much do about this after a bug report is send.
If there is not much interest in implementing new recommendations and to use them, there is not much use to think about new features ensuring more options or better accessibility for the audience.

Because EPUB can have a long lifetime, however it is useful to provide the option of alternative stylesheets as available for (X)HTML and SVG already now, that people can use it in the future, after the bugs in programs are fixed.
With a few programs/extensions one can test, that it really works as intended.
Maybe one can mention the functionality as an accessibility feature in the preface of a book, suggesting to use a more powerful viewer/reader, if the current one does not provide the functionality.
